### PR TITLE
New 3DS XL custom layout (issue Complete configuration options #4)

### PR DESCRIPTION
**Pull request by CDS83**
New 3DS XL custom layout proposal

### CODE REVIEW

File: src/citra_libretro/citra_libretro.cpp [hunk 1]
```diff
@@ -94,7 +94,8 @@ void LibRetro::OnConfigureEnvironment() {
         {"citra_use_acc_geo_shaders", "Enable accurate geometry shaders (only for H/W shaders); enabled|disabled"},
         {"citra_use_acc_mul", "Enable accurate shaders multiplication (only for H/W shaders); enabled|disabled"},
         {"citra_resolution_factor",
-         "Resolution scale factor; 1x (Native)|2x|3x|4x|5x|6x|7x|8x|9x|10x"},
+         "Resolution scale factor; 1x (Native)|2x|3x|4x|5x|6x|7x|8x|9x|10x|1x custom 1920/1080 (New 3DS XL)|2x custom 1920/1080 (New 3DS XL)|4x custom 1920/1080 (New 3DS XL)"},
+        {"citra_test", "TEST; 0"},
```
Comment (j-selby) [inline src/citra_libretro/citra_libretro.cpp]: What is this doing here?

File: CMakeLists.txt [hunk 2]
```diff
@@ -12,11 +12,11 @@ project(citra)
 option(ENABLE_SDL2 "Enable the SDL2 frontend" ON)
 CMAKE_DEPENDENT_OPTION(CITRA_USE_BUNDLED_SDL2 "Download bundled SDL2 binaries" ON "ENABLE_SDL2;MSVC" OFF)
 
-option(ENABLE_QT "Enable the Qt frontend" ON)
+option(ENABLE_QT "Enable the Qt frontend" OFF)
```
Comment (j-selby) [inline CMakeLists.txt]: This should not be changed - the point of this is that this repo by default generates builds that are compatible with upstream. This should be made in a different PR.

File: src/citra_libretro/citra_libretro.cpp [hunk 3]
```diff
@@ -194,6 +196,44 @@ void UpdateSettings() {
     } else {
         int scale = stoi(scaling.substr(0, endOfScale));
         Settings::values.resolution_factor = scale;
+        auto custom = scaling.find("custom");
+        if (custom != std::string::npos) {
+            switch(scale) {
+                case 1:
+                    Settings::values.custom_layout = true;
```
Comment (j-selby) [inline src/citra_libretro/citra_libretro.cpp]: What are these values specifically? Are they named?

File: src/citra_libretro/citra_libretro.cpp (hunk 1 above)
Comment (CDS83) [reply src/citra_libretro/citra_libretro.cpp]: Fixed... It seems i've been a bit too fast with this commit...

File: CMakeLists.txt (hunk 2 above)
Comment (CDS83) [reply CMakeLists.txt]: Same as above, you are right of course

File: src/citra_libretro/citra_libretro.cpp (hunk 3 above)
Comment (CDS83) [reply src/citra_libretro/citra_libretro.cpp]: The custom_xxx are defined in core/settings.h. If custom_layout is set to true, the others define the position and size of the two screens.
I'm not very happy with the hard-coded bunch of values, but I've not found another way to specify integers in such wide ranges (about 0-2000 or more I'd say). Other drawback : those values are screen resolution dependent. However with this option you have now a support for realistic 3DS overlays as in the screenshot in the issue.

Comment (j-selby) [reply src/citra_libretro/citra_libretro.cpp]: `Other drawback : those values are screen resolution dependent.`
imo I think this is a breaking issue. Can this instead be specified as percentages of the total screen res, and do operations on that?

This all should be documented in the code itself.

Comment (CDS83) [reply src/citra_libretro/citra_libretro.cpp]: Thanks for your feedback. I tried to dig into that direction by setting screens pos and size as ratio (per thousand, cause percentage is not precise enough) of the libretro window.
You are now free to move the screens inside the frame.
I've also defined a preset for 16/9 screens that works fine with an overlay I made (using overlays is the only scenario I can think of for such features).

Comment (CDS83) [reply src/citra_libretro/citra_libretro.cpp]: And I changed the name of "16/9 preset" to "New 3DS XL": this is more a matter of console proportion than screen ratio...

**Comment by dankcushions** [comment]
closed during hack-gate 2020. @twinaphex please re-open :)
